ddtrace/debugging/_session.py

```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass
+import typing as t
+from weakref import WeakKeyDictionary as wkdict
+
+from ddtrace import tracer
+
+
+@dataclass
+class Session:
+    ident: str
+    level: int
+
+    def link_to_trace(self, trace_context: t.Optional[t.Any] = None):
+        SessionManager.link_session_to_trace(self, trace_context)
+
+    @classmethod
+    def get_for_trace(cls) -> t.Optional["Session"]:
+        return SessionManager.get_session_for_trace()
+
+
+class SessionManager:
+    _session_trace_map: wkdict = wkdict()  # Trace context to Session mapping
+
+    @classmethod
+    def link_session_to_trace(cls, session, trace_context: t.Optional[t.Any] = None) -> None:
+        cls._session_trace_map[trace_context or tracer.current_root_span()] = session
+
+    @classmethod
+    def get_session_for_trace(cls) -> t.Optional[Session]:
```

Comment:
a trace can have multiple sessions assoiated to a trace.
we might want to have different API:
get_session_for_trace(cls, session_id) -> return a specific session (used by manual probes)
should_record_code_origin_snapshot() -> return true if one session has level > 2